src/libAtomVM/atom_table.c

```diff
@@ -0,0 +1,474 @@
+
+    SMP_RDLOCK(hash_table);
+    unsigned long bucket_index = hash % hash_table->capacity;
+    return get_node_from_bucket(hash_table, bucket_index, string);
+}
+
+long atom_table_get_index(struct AtomTable *table, AtomString string)
+{
+    struct HNode *node = lock_and_get_node(table, string);
+    long result = (node != NULL) ? node->index : ATOM_TABLE_NOT_FOUND;
+
+    SMP_UNLOCK(table);
+    return result;
+}
+
+// TODO: this function needs use an efficient structure such as a skip list
+static struct HNode *get_node_using_index(struct AtomTable *table, long index)
+{
+    if (UNLIKELY(index >= table->count)) {
+        return NULL;
+    }
+
+    struct HNodeGroup *node_group = table->first_node_group;
+    while (node_group) {
+        long first_index = node_group->first_index;
+        if (first_index + node_group->len > index) {
+            return &node_group->nodes[index - first_index];
+        }
+
+        node_group = node_group->next;
+    }
+
+    return NULL;
+}
+
+AtomString atom_table_get_atom_string(struct AtomTable *table, long index)
+{
+    SMP_RDLOCK(table);
+
+    struct HNode *node = get_node_using_index(table, index);
+    if (IS_NULL_PTR(node)) {
+        return NULL;
+    }
+
+    AtomString found_key = node->key;
+
+    SMP_UNLOCK(table);
+    return found_key;
+}
+
+int atom_table_cmp_using_atom_index(struct AtomTable *table, int t_atom_index, int other_atom_index)
+{
+    AtomString t_atom_string = atom_table_get_atom_string(table, t_atom_index);
+
+    int t_atom_len = atom_string_len(t_atom_string);
+    const char *t_atom_data = (const char *) atom_string_data(t_atom_string);
+
+    AtomString other_atom_string = atom_table_get_atom_string(table, other_atom_index);
+
+    int other_atom_len = atom_string_len(other_atom_string);
+    const char *other_atom_data = (const char *) atom_string_data(other_atom_string);
+
+    int cmp_size = (t_atom_len > other_atom_len) ? other_atom_len : t_atom_len;
+
+    int memcmp_result = memcmp(t_atom_data, other_atom_data, cmp_size);
+
+    if (memcmp_result == 0) {
+        if (t_atom_len == other_atom_len) {
+            return 0;
+        } else {
+            return (t_atom_len > other_atom_len) ? 1 : -1;
+        }
+    }
+
+    return memcmp_result;
+}
+
+atom_ref_t atom_table_get_atom_ptr_and_len(struct AtomTable *table, long index, size_t *out_len)
+{
+    SMP_RDLOCK(table);
+
+    struct HNode *node = get_node_using_index(table, index);
+    if (IS_NULL_PTR(node)) {
+        return NULL;
+    }
+
+    *out_len = atom_string_len(node->key);
+
+    SMP_UNLOCK(table);
+    return node;
+}
+
+void atom_table_write_bytes(struct AtomTable *table, atom_ref_t atom, size_t buf_len, void *outbuf)
+{
+    SMP_RDLOCK(table);
+
+    struct HNode *node = (struct HNode *) atom;
+    size_t len = atom_string_len(node->key);
+    if (len > buf_len) {
+        len = buf_len;
+    }
+
+    memcpy(outbuf, atom_string_data(node->key), len);
+
+    SMP_UNLOCK(table);
```

Comment:
Is the lock protecting the memory pointed to by node->key or could we put the memcpy outside?

Reply:
Yes, it is protecting it...

Or at least that's the plan on the longer term (we are still not able to ensure that because of `atom_table_get_atom_string` function).

As soon as we fix that, the contract will be that:
- `atom_table_ensure_atom` by default assumes that given atom string is in a constant area of memory (such as flash), so the table takes ownership of it.
- When `AtomTableCopyAtom` option is given, it means that a copy has to be performed because that memory is mutable and it cannot be owned
- It will be not possible to obtain any direct reference to any atom string. It will be allowed to get a copy into a caller owned buffer

Therefore we'll be able to replace atom strings when doing module load/unload and stuff like that.